File: services/claims_scoring/src/checks/frequency.rs
use anyhow::Result;
use sqlx::PgPool;
use tracing::{error, info};
use uuid::Uuid;

/// Result of the 90-day frequency check
#[derive(Debug, Clone)]
pub struct FrequencyResult {
    /// Frequency sub-score: 1.0 if claims_90d <= 3, 0.0 if > 3
    pub score: f64,
    /// True if the worker has exceeded the 90-day claim velocity cap (>3 approved claims)
    pub velocity_cap_exceeded: bool,
    /// Number of approved claims in the prior 90-day window
    pub claims_90d: i64,
}

/// Checks how many approved claims the worker has submitted in the prior 90-day window.
///
/// - claims_90d <= 3 → score 1.0, velocity_cap_exceeded = false
/// - claims_90d > 3  → score 0.0, velocity_cap_exceeded = true (routes to FRAUD_QUEUE)
///
/// On DB error, returns a conservative score of 0.5 and logs the error.
pub async fn check_frequency(pool: &PgPool, worker_id: Uuid) -> Result<FrequencyResult> {
    let result = run_frequency_query(pool, worker_id).await;

    match result {
        Ok(r) => Ok(r),
        Err(e) => {
            error!(
                error = %e,
                %worker_id,
                "PostgreSQL frequency query failed — using conservative default score 0.5"
            );
            Ok(FrequencyResult {
                score: 0.5,
                velocity_cap_exceeded: false,
                claims_90d: 0,
            })
        }
    }
}

async fn run_frequency_query(pool: &PgPool, worker_id: Uuid) -> Result<FrequencyResult> {
    let count: i64 = sqlx::query_scalar(
        r#"
        SELECT COUNT(*)
        FROM claims
        WHERE worker_id = $1
          AND status IN ('auto_approved', 'approved')
          AND submitted_at > NOW() - INTERVAL '90 days'
        "#,
    )
    .bind(worker_id)
    .fetch_one(pool)
    .await?;

    let claims_90d = count;
    let velocity_cap_exceeded = claims_90d > 3;
    let score = if velocity_cap_exceeded { 0.0 } else { 1.0 };

    info!(
        %worker_id,
        claims_90d,
        velocity_cap_exceeded,
        "Frequency check complete"
    );

    Ok(FrequencyResult {
        score,
        velocity_cap_exceeded,
        claims_90d,
    })
}

File: services/claims_scoring/checks/frequency.py
import logging
from dataclasses import dataclass

import asyncpg

logger = logging.getLogger(__name__)


@dataclass
class FrequencyResult:
    score: float
    velocity_cap_exceeded: bool
    claims_90d: int


async def check_frequency(pool: asyncpg.Pool, worker_id: str) -> FrequencyResult:
    try:
        return await _run_frequency_query(pool, worker_id)
    except Exception as e:
        logger.error(
            "PostgreSQL frequency query failed worker_id=%s error=%s — using score 0.5",
            worker_id, e,
        )
        return FrequencyResult(score=0.5, velocity_cap_exceeded=False, claims_90d=0)


async def _run_frequency_query(pool: asyncpg.Pool, worker_id: str) -> FrequencyResult:
    count = await pool.fetchval(
        """
        SELECT COUNT(*)
        FROM claims
        WHERE worker_id = $1
          AND status IN ('auto_approved', 'approved')
          AND submitted_at > NOW() - INTERVAL '90 days'
        """,
        str(worker_id),
    )
    claims_90d = int(count or 0)
    velocity_cap_exceeded = claims_90d > 3
    score = 0.0 if velocity_cap_exceeded else 1.0

    logger.info(
        "Frequency check worker_id=%s claims_90d=%d velocity_cap_exceeded=%s",
        worker_id, claims_90d, velocity_cap_exceeded,
    )

    return FrequencyResult(score=score, velocity_cap_exceeded=velocity_cap_exceeded, claims_90d=claims_90d)
